h2o-core/src/main/java/water/FailedNodeWatchdogExtension.java

```diff
@@ -0,0 +1,155 @@
+            if (s.matches("watchdog_client_retry_timeout")) {
+                watchdogClientRetryTimeout = s.parseInt(args[i + 1]);
+                String[] new_args = new String[args.length - 2];
+                System.arraycopy(args, 0, new_args, 0, i);
+                System.arraycopy(args, i + 2, new_args, i, args.length - (i + 2));
+                return new_args;
+            }
+            else if(s.matches("watchdog_client")){
+                watchDogClient = true; H2O.ARGS.client = true;
+                String[] new_args = new String[args.length - 1];
+                System.arraycopy(args, 0, new_args, 0, i);
+                System.arraycopy(args, i + 1, new_args, i, args.length - (i + 1));
+                return new_args;
+            }
+        }
+        return args;
+    }
+
+
+    public void validateArguments() {
+        if (watchdogClientRetryTimeout < 0) {
+            H2O.parseFailed("Watchdog client retry timeout has to be positive: " + watchdogClientRetryTimeout);
+        }
+    }
+
+    @Override
+    public void onLocalNodeStarted() {
+        new FailedNodeWatchdogThread().start();
+        H2O.SELF._heartbeat._watchdog_client = watchDogClient;
+    }
+
+
+    /**
+     * This method checks whether the client is disconnected from this node due to some problem such as client or network
+     * is unreachable.
+     */
+    private static void handleClientDisconnect(H2ONode node) {
+        if(node._heartbeat._watchdog_client){
+            Log.warn("Watchdog client " + node + " disconnected!");
+            BullyClientGoneTask tsk = new BullyClientGoneTask(node);
+            Log.warn("Asking the rest of the nodes in the cloud whether watchdog client is really gone.");
+            if(((BullyClientGoneTask)tsk.doAllNodes()).consensus) {
+                Log.fatal("Stopping H2O cloud since the watchdog client is disconnected from all nodes in the cluster!");
+                H2O.shutdown(0);
+            }
+        }else if(node._heartbeat._client) {
+            Log.warn("Client "+ node +" disconnected!");
+        }
+
+        // in both cases remove the client
+        if(node._heartbeat._client){
+            H2O.removeClient(node);
+            if(H2O.isFlatfileEnabled()){
+                H2O.removeNodeFromFlatfile(node);
+            }
+        }
+    }
+
+    /**
+     * Helper MR task used to detect consensus on the fact that bully client is disconnected from the network
+     */
+    private static class BullyClientGoneTask extends MRTask {
+        public boolean consensus = true;
+        private H2ONode clientNode;
+
+        public BullyClientGoneTask(H2ONode clientNode) {
+            this.clientNode = clientNode;
+        }
+
+        @Override
+        public void reduce(MRTask mrt) {
+            if(((BullyClientGoneTask)mrt).consensus) { // don't change the value in case negative consensus
+                for(H2ONode node: H2O.getClients()){
```

Comment:
Oh I see the logic is in reduce. That is not gonna work, reduce is not run on the leaf nodes. You want to move this into setupLocal. Reduce should only do this.consensus = this.consensus && other.consensus. (Also by extending MRTask<BullyGoneTsk> you can get remove of the typecast. Just cosmetics but it looks nicer : )